tools_misc/wrl_to_obj.py

```diff
@@ -0,0 +1,148 @@
+            # correspond to the coordinates above, or faces.
+            [ 'MaterialBinding' '{' 'value' mtl_binding:binding_type '}' ]
+
+            # The material has two sub-sections whose lengths are not guaranteed
+            # to match. The diffuseColor array contains a number of float
+            # triplets (RGB) equal to the number of bound elements. However, in
+            # the case of PER_FACE bound materials, the transparency array has
+            # a single float rather than one per diffuseColor.
+            [
+                'Material'
+                '{'
+                    'diffuseColor' '[' mtl_diffuse_colors:points ']'
+                    'transparency[' mtl_transparencies:floats ']'
+                '}'
+            ]
+
+            # An indexed line set contains 1 or more edges, each defined by a
+            # pair of indexes into the coords array.
+            [
+                'IndexedLineSet'
+                '{'
+                    'coordIndex' '[' indexed_lines:index_groups ']'
+                '}'
+            ]
+
+            # Similarly to the indexed lines, the faces are triplets of indexes.
+            [
+                'IndexedFaceSet'
+                '{'
+                    'coordIndex' '[' indexed_faces:index_groups ']'
+                '}'
+            ]
+        '}'
+        ;
+
+    # If the binding type is PER_VERTEX, then the material properties map to the
+    # coordinates array. For PER_FACE, they map to the IndexedFaceSet.
+    binding_type = 'PER_VERTEX' | 'PER_FACE' ;
+
+    # Index arrays appear "flat", but actually use -1 as a terminator marking
+    # the end of spans of indexes within. There's never actually any negative
+    # indexes. Also note that the indexes may sometimes be output in a single
+    # line, or when there's many values they can be output over multiple lines
+    # in which case they will also gain a trailing comma before the closing "]".
+    index_groups = { @+:index_group ',' } [ @+:index_group ] ;
+    index_group = { @+:index ',' }+ '-1' ;
+    index = /\\d+/ ;
+
+    # Within an array, points are separated by comma (optional trailing comma)
+    points = { @+:point ',' } [ @+:point ] ;
+    # Points/colours are float triplets separated by whitespace.
+    point = x:float y:float z:float ;
+
+    # The only places where a comma-separated list of single floats is seen is
+    # the transparency material property.
+    floats = { @+:float ',' } [ @+:float ] ;
+
+    # Floats can be negative and always have a 6 digit decimal part
+    float = /-?\\d+\\.\\d+/ ;
+""")
+
+# Given a string of the WRL content, returns its abstract syntax tree.
+def parse_wrl_to_ast(wrl_content):
+    return WRL_PARSER.parse(wrl_content)
+
+# Translates the WRL input stream to an OBJ output stream.
+# The caller is responsible for closing streams.
+def convert_wrl_to_obj(input_stream, output_stream):
+    def output(line):
+        print(line, file=output_stream)
```

Comment:
Python allows for documentation to stored in the first statement of a function/object definition.

Done as follows:
```py
def parse_wrl_to_ast(wrl_content):
    '''
    Given a string of the WRL content, returns its abstract syntax tree.
    '''
    return WRL_PARSER.parse(wrl_content)
```

This makes it so that whenever automatic documentation generators are ran over projects that function description actually ends up in the documentation.